File: backend/api/v1/endpoints/pricing.py
"""Pricing and Greeks endpoints."""
from fastapi import APIRouter, HTTPException
from backend.schemas.pricing import (
    PricingRequest,
    PricingResponse,
    GreeksRequest,
    GreeksResponse,
)
from backend.services.pricing_service import PricingService

router = APIRouter()
pricing_service = PricingService()


@router.post("/price", response_model=PricingResponse, tags=["Pricing"])
async def calculate_price(request: PricingRequest):
    """
    Calculate option price using Black-Scholes model.

    Returns the theoretical price for a European-style option.
    """
    try:
        return pricing_service.calculate_price(request)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/greeks", response_model=GreeksResponse, tags=["Pricing"])
async def calculate_greeks(request: GreeksRequest):
    """
    Calculate all Greeks (delta, gamma, vega, theta, rho) for an option.

    Greeks measure the sensitivity of option price to various factors:
    - Delta: Sensitivity to underlying price
    - Gamma: Rate of change of delta
    - Vega: Sensitivity to volatility
    - Theta: Time decay (per day)
    - Rho: Sensitivity to interest rate
    """
    try:
        return pricing_service.calculate_greeks(request)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/types/pricing.ts
/**
 * Pricing and Greeks types
 */

export type OptionType = 'call' | 'put';

export interface PricingRequest {
  spot: number;
  strike: number;
  time_to_expiry: number;
  volatility: number;
  risk_free_rate: number;
  dividend_yield: number;
  option_type: OptionType;
}

export interface PricingResponse {
  price: number;
  model: string;
}

export interface GreeksRequest {
  spot: number;
  strike: number;
  time_to_expiry: number;
  volatility: number;
  risk_free_rate: number;
  dividend_yield: number;
  option_type: OptionType;
}

export interface GreeksResponse {
  delta: number;
  gamma: number;
  vega: number;
  theta: number;
  rho: number;
}
